### ISSUE TITLE
Talos scans and custom models showing very different performance

### ISSUE THREAD

**Issue by vascosa** [autonomio/talos#87]
Hello,

I don't quite understand how Talos is optimizing the Keras models it produces and as such I'm having trouble replicating the results outside of Talos. 

I build a Talos model with different parameters did 14 scans, a 0.01 grid downsample. Accuracy results ranged from 68% to 89% (mean 81%), and fmeasures ranging from 0.25 to 0.86 (mean 0.67). The best measures results from 

```
p = {'first_neuron':[ 8],
     'hidden_layers':[2],
     'batch_size': [30],
     'epochs': [200],
     'dropout': [00.5],
     'kernel_initializer': ['normal'],
     'shape':['brick'],
     'optimizer': [Nadam],
     'losses': [sparse_categorical_crossentropy],
     'activation':[relu],
     'last_activation': [sigmoid]}
```

Then I tried to build a model to replicate this on the same data:
```
X, Y = import_data()
X = preprocessing.scale(X)
x, X_test, y, y_test = train_test_split(X, Y, test_size = 0.2, random_state=44)

best_model=Sequential()
best_model.add(Dense(8, activation='relu', input_dim=X_test.shape[1]))
best_model.add(Dropout(0.5))
best_model.add(Dense(8, activation='relu'))
best_model.add(Dense(8, activation='relu'))
best_model.add(Dense(2, activation='sigmoid'))

best_model.compile(loss='sparse_categorical_crossentropy',
                      optimizer='Nadam',
                      metrics=['accuracy',fmeasure])

best_model.fit(x, y, epochs=200, batch_size=30)
score = best_model.evaluate(X_test, y_test, verbose=0)
print(score)
```

This model never surpasses  80% accuracy or 0. 7fmeasure. Neither does the model created based on the worst scan. For both models I got accuracies as low as 50%, meaning the range is really different from what that of the scans indicate. I would expect at least one scan to be low as well, but that's not the case. What is causing the difference in performance between my test model and the Talos model?

For reference, this is my Talos model:

```
def newmodel(x_train, y_train, x_val, y_val, params):

    model = Sequential()
    model.add(Dense(params['first_neuron'], input_dim=x_train.shape[1],
                    activation=params['activation'],
                    kernel_initializer=params['kernel_initializer']))    
    model.add(Dropout(params['dropout'])) 
    hidden_layers(model, params, 1)
    model.add(Dense(2, activation=params['last_activation'],
                    kernel_initializer=params['kernel_initializer']))  
   model.compile(loss=params['losses'],
                  optimizer=params['optimizer'](),
                  metrics=['acc', fmeasure])  
    history = model.fit(x_train, y_train, 
                        validation_data=[x_val, y_val],
                        batch_size=params['batch_size'],
                        epochs=params['epochs'],
                        verbose=0)

    return history, model


p = {'first_neuron':[4, 8, 16, 32, 64],
     'hidden_layers':[0, 1, 2],
     'batch_size': [2, 30],
     'epochs': [50, 200],
     'dropout': [0.2, 0.5],
     'kernel_initializer': ['normal'],
     'shape':['brick','long_funnel'],
     'optimizer': [Nadam, Adam, RMSprop],
     'losses': [sparse_categorical_crossentropy],
     'activation':[relu, elu],
     'last_activation': [sigmoid]}


t = ta.Scan(x=X,
            y=Y,
            model=newmodel,
            grid_downsample=0.01, 
            params=p,
            dataset_name='LA5s',
            val_split = 0.2,
            save_best_model = True,
            experiment_no='1')
```

Thank you so much for your help.

Best,
Vasco

**Comment by x94carbone** [comment]
Could you be more specific about how you evaluate the model? The values that Talos reports are of course on the validation data, so the final reported value will be worse since you should be reporting only the values evaluated on the training set.

In other words, your first results, where you took the highest validation data, might not give you as good of a score if you’re reporting the testing dataset value. This could just mean your model is not generalizing well.

To clarify, could you specify exactly which datasets you’re reporting on? My understanding is that the first accuracies you’re reporting are on the validation data and the second on testing data? I only wish to clarify because I have not had this specific problem with Talos in my useage of it. Seems like either an overfitting or a data problem. Regardless I’ll try to help you get to the bottom of it 👍

**Comment by vascosa** [comment]
Thanks for the quick reply. What I am am trying to do outside of Talos is replicating what Talos is doing but manually. I am currently working with about 10% of my data (derived from fMRI) as I tweak the preprocessing of the data. I used the full 10% as input for talos (with the 80/20 split). I used the exact same data outside of Talos. The only reason I'm doing this is because the sample is small and I'm just making sure things are more or less working for when I have the actual full dataset ready.

Today I ran a simple test where I had Talos randomly search for 0.01 scans and give me the results. Then, I replicated the 14 scans manually  with a train_test_split (80/20) in a similar fashion to what Talos does.  The code I used is all above, all I changed were the model variables. When I import the data `X, Y = import_data()` I'm importing the exact same data for Talos and for my manual tests.

The results are attached. I added two columns to the Talos csv output: manual validation accuracy; and manual validation f measure. It seams like the fmeasure doesn't change significantly, but the accuracy is quite a bit lower. I would expect, just by chance, that these results would not differ, as I'm just manually replicating what Talos is doing automatically (supposedly) - I think the issue might be with the way I'm replicating the models/scans manually.

Let me know and thanks for your help!


[LA5s_1.xlsx](https://github.com/autonomio/talos/files/2348376/LA5s_1.xlsx)

**Comment by x94carbone** [comment]
`I think the issue might be with the way I'm replicating the models/scans manually.`

It very well might be. Additionally, looking at your final loss and accuracy, it appears to me that you're over-fitting quite a bit. If this is true, then it is unlikely you're going to be able to produce reproducible results upon training since each time you train you'll be tailoring your model to the training data or the model itself. Sometimes you'll get lucky and have similar-enough validation data, and other times you won't, hence the large variance.

Here's my recommendation to help you pin down the potential problem with Talos first. First, do a three-way stratified (equal proportion of each class in each chunk) split. One for training (~80%) one for validation (~10%) one for testing (~10%). Manually specify the input validation data into Talos so you know exactly how your data is being split. Run the hyper-parameter tuning on a reproducible chunk of your hyperparameter space (meaning, specify in the parameter variable - the input to Talos - only the hyperparameter permutations you want and do not use any reduction of the hyperparameter space). See what the results are like. Then, pick the best model, retrain on those hyperparameters with only the training data, and finally evaluate on the testing data (that should be the end-all result you report). To confirm everything is working properly, then you can try manually reproducing everything I just mentioned. See if there results are similar!

Also, I saw you're using scaling of some sort, but you're scaling the entire data set. Might I ask what the scaler does exactly? Some scalers are only meant to be "fit" on the training data, else you risk introducing additional biases.

I would try to address the over-fitting problem first though, but based on your model you already don't have many parameters to train. You said this was fMRI data. Is this a 2/3d image that has been flattened? If so you might want to consider using a CNN, it could help you generalize a bit better. Sorry if you've thought of these things already, I'm just brainstorming! Not trying to waste your time or anything.

Let me know what you think!

**Comment by mikkokotila** [comment]
Hello @vascosa. In addition to what @x94carbone had shared already, it looks like your model has some differences with the Talos experiment parameters:

- dropout is missing between hidden layers
- shape might be different (at least long_funnel is quite different from 8-8-8-2)
- number of input neurons is different
- kernel initializer is missing

**Comment by mikkokotila** [comment]
I guess this was resolved so closing here. Feel free to reopen if more information.
